solidago/src/solidago/scoring_model.py

```diff
@@ -0,0 +1,217 @@
+        else:
+            right_uncertainty += (- base_right) * min(
+                self.multiplicator_left_uncertainty, self.multiplicator)
+            
+        return score, left_uncertainty, right_uncertainty
+        
+    def scored_entities(self, entities=None) -> set[int]:
+        return self.base_model.scored_entities(entities)
+
+    def _direct_scaling_parameters(self):
+        return (self.multiplicator, self.translation, 
+            self.multiplicator_left_uncertainty, self.multiplicator_right_uncertainty, 
+            self.translation_left_uncertainty, self.translation_right_uncertainty)
+
+    def get_scaling_parameters(self):
+        model, parameters = self, list()
+        while hasattr(model, "base_model"):
+            model = model.base_model
+            parameters.append(model._direct_scaling_parameters())
+        return ScaledScoringModel.compose_scaling_parameters(parameters)
+
+    def __str__(self, indent=""):
+        result = indent + "{\n" 
+        result += f"{indent}    multiplicator = {np.round(self.multiplicator, 2)}"
+        result += f"   [{np.round(self.multiplicator_left_uncertainty, 2)}, "
+        result += f"{np.round(self.multiplicator_right_uncertainty, 2)}]\n{indent}"
+        result += f"    translation = {np.round(self.translation, 2)}"
+        result += f"   [{np.round(self.translation_left_uncertainty, 2)}, "
+        result += f"{np.round(self.translation_right_uncertainty, 2)}]\n{indent}"
+        result += "    base_model = " + self.base_model.__str__(f"    {indent}")
+        result += "\n" + indent + "}"
+        return result
+
+    @classmethod
+    def compose_scaling_parameters(parameters):
+        result = 1, 0, 0, 0, 0, 0
+        for m2, t2, m2_left, m2_right, t2_left, t2_right in parameters:
+            m1, t1, m1_left, m1_right, t1_left, t1_right = result
+            result[0] = m1 * m2
+            result[1] = t2 * m1 + t1
+            result[2] = m1_left  * m2 + m2_left  * m1
+            result[3] = m1_right * m2 + m2_right * m1
+            result[4] = t2_left  * m1 + t1_left
+            result[5] = t2_right * m1 + t1_right
+        return result
+
+
+class PostProcessedScoringModel(ScoringModel):
+    def __init__(self, base_model: ScoringModel, post_process: callable):
+        """ Defines a derived scoring model, based on a base model and a post process
+        
+        Parameters
+        ----------
+        base_model: ScoringModel
+        post_process: callable
+            Must be a monotonous function float -> float
+        """
+        self.base_model = base_model
+        self.post_process = post_process
+    
+    def __call__(self, entity_id, entity_features):
+        base_score, base_left, base_right = self.base_model(entity_id, entity_features)
+        score = self.post_process(base_score)
+        left = self.post_process(base_score - base_left) - score
+        right = self.post_process(base_score - base_right) - score
```

Comment:
@lenhoanglnh Shouldn't we evaluate the post_process function on the right of the base_score too?
Intuitively, I would have expected something like:
```suggestion
        left = score - self.post_process(base_score - base_left)
        right = self.post_process(base_score + base_right) - score
```